nixos/modules/services/matrix/mjolnir.nix

```diff
@@ -0,0 +1,239 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.mjolnir;
+
+  yamlConfig = {
+    inherit (cfg) dataPath managementRoom protectedRooms;
+
+    accessToken = "@ACCESS_TOKEN@"; # will be replaced in "generateConfig"
+    homeserverUrl =
+      if cfg.pantalaimon.enable then
+        "http://${cfg.pantalaimon.options.listenAddress}:${toString cfg.pantalaimon.options.listenPort}"
+      else
+        cfg.homeserverUrl;
+
+    pantalaimon = {
+      inherit (cfg.pantalaimon) username;
+
+      use = cfg.pantalaimon.enable;
+      password = "@PANTALAIMON_PASSWORD@"; # will be replaced in "generateConfig"
+    };
+  };
+
+  moduleConfigFile = pkgs.writeText "module-config.yaml" (
+    generators.toYAML { } (filterAttrs (_: v: v != null)
+      (fold recursiveUpdate { } [ yamlConfig cfg.settings ])));
+
+  # these config files will be merged one after the other to build the final config
+  configFiles = [
+    "${pkgs.mjolnir}/share/mjolnir/config/default.yaml"
+    moduleConfigFile
+  ];
+
+  # this will generate the default.yaml file with all configFiles as inputs and
+  # replace all secret strings using replace-secret
+  generateConfig = pkgs.writeShellScript "mjolnir-generate-config" (
+    let
+      yqEvalStr = concatImapStringsSep " * " (pos: _: "select(fileIndex == ${toString (pos - 1)})") configFiles;
+      yqEvalArgs = concatStringsSep " " configFiles;
+    in
+    ''
+      # mjolnir will try to load a config from "./config/default.yaml" in the working directory
+      # -> let's place the generated config there
+      mkdir -p ${cfg.dataPath}/config
```

Comment:
It looks like the mjolnir config is created with some somewhat generous permissions:

```
[root@kif:~]# stat /var/lib/mjolnir/config/default.yaml
  File: /var/lib/mjolnir/config/default.yaml
  Size: 6550      	Blocks: 10         IO Block: 6656   regular file
Device: 1bh/27d	Inode: 49088       Links: 1
Access: (0644/-rw-r--r--)  Uid: (  990/ mjolnir)   Gid: (  989/ mjolnir)
Access: 2021-06-14 15:07:38.557290877 +0000
Modify: 2021-06-14 15:12:39.376004070 +0000
Change: 2021-06-14 15:12:39.376004070 +0000
 Birth: -

```

however an unprivileged user can't access it due to the parent directory:

```
[root@kif:~]# stat /var/lib/mjolnir/
  File: /var/lib/mjolnir/
  Size: 4         	Blocks: 1          IO Block: 131072 directory
Device: 1bh/27d	Inode: 264677      Links: 3
Access: (0700/drwx------)  Uid: (  990/ mjolnir)   Gid: (  989/ mjolnir)
Access: 2021-06-14 15:07:38.538291044 +0000
Modify: 2021-06-14 15:12:02.775251291 +0000
Change: 2021-06-14 15:12:02.775251291 +0000
 Birth: -
```

but I wonder if we should drop the umask anyway.

Reply:
added an more restrictive `umask`